apps/game_backend/lib/game_backend/units.ex

```diff
@@ -305,4 +308,92 @@ defmodule GameBackend.Units do
     |> UnitSkin.changeset(attrs)
     |> Repo.insert()
   end
+
+  @doc """
+  Levels up a user's unit and substracts the currency cost from the user.
+
+  Returns `{:error, :not_found}` if unit doesn't exist or if it's not owned by user.
+  Returns `{:error, :cant_afford}` if user cannot afford the cost.
+  Returns `{:ok, %{unit: %Unit{}, user_currency: %UserCurrency{}}}` if succesful.
+  """
+  def level_up(user_id, unit_id) do
+    with {:unit, {:ok, unit}} <- {:unit, get_unit(unit_id)},
+         {:unit_owned, true} <- {:unit_owned, unit.user_id == user_id},
+         level_up_config = get_current_level_up_config(),
+         {:can_level_up, true} <-
+           {:can_level_up,
+            Enum.any?(level_up_config.level_info, fn level_info -> level_info.level == unit.level + 1 end)},
+         costs = calculate_level_up_cost(unit, level_up_config),
+         {:can_afford, true} <-
+           {:can_afford, Currencies.can_afford(user_id, costs)} do
+      result =
+        Multi.new()
+        |> Multi.run(:unit, fn _, _ -> add_level(unit) end)
+        |> Multi.run(:user_currency, fn _, _ ->
+          Currencies.substract_currencies(user_id, costs)
+        end)
+        |> Transaction.run()
+
+      case result do
+        {:error, reason} ->
+          {:error, reason}
+
+        {:error, _, _, _} ->
+          {:error, :transaction}
+
+        {:ok, %{unit: unit, user_currency: user_currency}} ->
+          {:ok, %{unit: unit, user_currency: user_currency}}
+      end
+    else
+      {:unit, {:error, :not_found}} -> {:error, :not_found}
+      {:can_level_up, false} -> {:error, :no_more_levels}
+      {:unit_owned, false} -> {:error, :not_found}
+      {:can_afford, false} -> {:error, :cant_afford}
+    end
```

Comment:
This is pretty hard to follow. I can't see what's going on here.
Please refactor this to be like other controllers. A `with` `case` and `else` it's a worse case than the antipattern showed here: https://hexdocs.pm/elixir/code-anti-patterns.html#complex-else-clauses-in-with